File: practice/binary_tree/binary_tree_level_order_traversal.py
# binary_tree_level_order_traversal.py --- 
# 
# Filename: binary_tree_level_order_traversal.py
# Description: 
# 
# Author:    Yu Lu
# Email:     yulu@utexas.edu
# Github:    https://github.com/SuperYuLu 
# 
# Created: Mon Oct 22 00:29:14 2018 (-0500)
# Version: 
# Last-Updated: Mon Oct 22 00:30:12 2018 (-0500)
#           By: yulu
#     Update #: 2
# 


class Solution:
    '''
    Given a binary tree, return the level order traversal of its nodes' values. (ie, from left to right, level by level).

    Definition for a binary tree node.
    class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

    '''
    
    # NOT an efficient solution, only beats 10% submissions 
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        res = []
        nextQueue = [root] if root else []
        while nextQueue:
            queue = nextQueue
            nextQueue = []
            currLayer = []
            while queue:
                curr = queue.pop()
                currLayer.append(curr.val)
                if curr.left:
                    nextQueue.insert(0, curr.left)
                if curr.right:
                    nextQueue.insert(0, curr.right)
            res.append(currLayer)        
        return res 
            
        

File: practice/binary_tree/binary_tree_level_order_traversal.cc
// level.cc --- 
// 
// Filename: level.cc
// Description: 
// 
// Author:    Yu Lu
// Email:     yulu@utexas.edu
// Github:    https://github.com/SuperYuLu 
// 
// Created: Thu Nov 22 15:33:55 2018 (-0600)
// Version: 
// Last-Updated: Thu Nov 22 15:34:05 2018 (-0600)
//           By: yulu
//     Update #: 1
// 

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  vector<vector<int>> levelOrder(TreeNode* root) {
    queue<TreeNode*> nodeQueue;
    queue<TreeNode*> nextQueue;
    vector<vector<int>> res;
    vector<int> levelRes;
    TreeNode* node;
    if(root) nodeQueue.push(root);
        
    while(!nodeQueue.empty() || !nextQueue.empty())
      {   
	if(nodeQueue.empty()) // move to next level
	  {
	    nodeQueue = nextQueue;
	    nextQueue = {};
	    res.push_back(levelRes);
	    levelRes = {};
	  }
	node = nodeQueue.front();
	nodeQueue.pop();
	//cout << node -> val << endl;
	levelRes.push_back(node -> val);
	if (node -> left) nextQueue.push(node -> left);
	if (node -> right) nextQueue.push(node -> right);
      }
    if(!levelRes.empty()) res.push_back(levelRes); // add last level result 
    return res;
        
  }
};

        
